compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,602 @@
+
+  collectSymbolSetSymExprVec(fn, symSet, symExprs);
+  buildDefUseMaps(symSet, symExprs, defMap, useMap);
+
+  //data structures for deferred actual handling
+  std::set<CallExpr*> deferredFns;
+  ActualUseDefCastMap actualUseDefMap;
+
+  bool cachedGlobalManip = globalManipFuncCache.count(fn);
+
+  forv_Vec(Symbol, sym, symSet) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar;
+    Type* castTo = NULL;
+
+    //if we don't already have it cached,
+    //check for global symbols in function body
+    if(!cachedGlobalManip) {
+      //I am not sure if there should be !isConstant and !isParam as well
+      //I think there can be a const but stateful global that is manipulated by
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
+        globalManipFuncCache[fn] = true;
+        cachedGlobalManip = true;
+      }
+    }
+
+    if(isDenormalizable(sym, defMap, useMap, &use, &def, &castTo)) {
+      usePar = use->parentExpr;
+      defPar = def->parentExpr;
+
+      //defer if the symbol used as actual
+      if(CallExpr* useCe = toCallExpr(usePar)){
+        if(!useCe->isPrimitive()) {
+          deferredFns.insert(useCe);
+          std::pair<Expr*, Type*> defCastBundle(def, castTo);
+          actualUseDefMap.insert(std::pair<SymExpr*, std::pair<Expr*, Type*> >
+              (use, defCastBundle));
+          continue;
+        }
+      }
+
+      //denormalize if the def is safe to move and there is no unsafe function
+      //between use and def
+      if(isExprSafeForReorder(def)) {
+        if(!possibleDepInBetween(defPar, usePar)) {
+          denormalize(def, use, castTo);
+        }
+      }
+    }
+  } // end loop for symbol
+
+  if(!cachedGlobalManip) {
+    globalManipFuncCache[fn] = false;
+  }
+
+  //handle deferred actuals
+  for(std::set<CallExpr*>::iterator ceIt = deferredFns.begin() ;
+      ceIt != deferredFns.end() ; ceIt++) {
+    denormalizeActuals(*ceIt, defMap, useMap, actualUseDefMap);
+  }
+
+  freeDefUseMaps(defMap, useMap);
+}
+
+void denormalizeActuals(CallExpr* ce,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap,
+    ActualUseDefCastMap actualUseDefMap) {
+
+  INT_ASSERT(!ce->isPrimitive());
+  for_alist_backward(actual, ce->argList) {
+    if(SymExpr* argSym = toSymExpr(actual)) { //else it's already denormd
+      if(! (argSym->var->isConstant() || argSym->var->isParameter())) {
+        if(actualUseDefMap.count(argSym) > 0) {
+          std::pair<Expr*,Type*> tmpTuple = actualUseDefMap[argSym];
+          SymExpr* use = argSym;
+          Expr* def = tmpTuple.first;
+          Type* castTo = tmpTuple.second;
+          Expr* usePar = use->parentExpr;
+          Expr* defPar = def->parentExpr;
+          if(CallExpr* ceTmp = toCallExpr(defPar)) {
+            if(!isRecord(ceTmp->get(1)->typeInfo())) { //to preserve pass-by-value
+              if(isExprSafeForReorder(def)) {
```

Comment:
Why does isExprSafeForReorder exist at all? If I had

      var a = someSideEffectingCall();  // e.g. call contains writeln()
      someOtherFunction(a);

then it would be valid to change it to

      someOtherFunction(someSideEffectingCall());

because the C compiler must run someSideEffectingCall() before someOtherFunction().

So, to me, it seems that there is only 1 issue: could the denormalization change the order of execution of the calls?

Consider this example:

      var a = someSideEffectingCall();  // e.g. call contains writeln()
      var b = someSideEffectingCall();  // e.g. call contains writeln()
      someOtherFunction(a,b);

We can change that into

      var a = someSideEffectingCall();  // e.g. call contains writeln()
      someOtherFunction(a, someSideEffectingCall());

without changing the order of the side-effects.

I'll use the terms "reads state" and "modifies state". Generally a function with "side effects" would "modify state". E.g. doing I/O or writing to global or local variables.

So, I'd use these rules for denormalizing:

 * If the definition and the use have any control flow or operations that modify state between them, do not denormalize
 * Otherwise, if the use is the last actual in a function call, it can be denomralized
 * Otherwise it cannot be denormalized

The only way I can think of improving upon these rules is to account for some operations that can always be reordered (and I suspect PRIM_ADDR_OF is that way) but I don't think that improvement is critical right now.